### DISCUSSION TITLE
SR code improvement and support for more than 4 axis

### DISCUSSION THREAD

**Discussion by Volksolive** [grblHAL/RP2040#23] [Q&A]
@terjeio, while I was looking at the PIO code for SR step&dir, I can see that it is not very efficient. I have improve it so that it used way less PIO memory and also one less SM. 
And as a bonus it can now support up to 8 axis and no dir->step delay when no dir changes.

I am still working on it and I haven't tested it so far but I am quite confident it can work.

I have a question though on the actual code for the 4 axis, why in driver.c, is it called M3 and not A? 
So now adding more axis what should be the name of these extra axis if not A, B, C, U, V?

I am putting the code here if you're interested before a future pull request (if everything works as planned):

![image](https://user-images.githubusercontent.com/78042392/151083447-bee60369-0b97-4c4c-8919-9b1c4c20b919.png)

![image](https://user-images.githubusercontent.com/78042392/151083484-be235114-29c8-4e54-9ad1-4a73aca38ff2.png)

BTW, there is now a PIO asm syntax highlighter available for VSCode.

**Comment by terjeio** [comment]
> And as a bonus it can now support up to 8 axis and no dir->step delay when no dir changes.

This will introduce a bit of jitter on dir changes...

> I have a question though on the actual code for the 4 axis, why in driver.c, is it called M3 and not A?

Because the fourth motor (M3) can be assigned to A, ganged X, ganged Y or ganged Z depending on the configuration.

> So now adding more axis what should be the name of these extra axis if not A, B, C, U, V?

M3 ... Mn, since they are motor designations - not axis designations. X, Y and Z are both.

EDIT: fixed the line above.

**Reply by Volksolive** [reply]
> This will introduce a bit of jitter on dir changes... 

Ok, will remove it !

> M3 ... Mn, since they are motor designations - not axis designations. A, B and C are both.

Should then everything be Mn like X Y Z too ? I just find it strange that we start with XYZ and then switch to M3, M4, Mn, etc.

**Reply by terjeio** [reply]
> Should then everything be Mn like X Y Z too ? I just find it strange that we start with XYZ and then switch to M3, M4, Mn, etc.

That would perhaps be ideal - but I am not going to spend the time needed to modify and reverify all the drivers. M3... etc. was introduced to solve the problem of configuring ganged axes vs ABC axes - and I only had to modify the few drivers that supported ganging or more than 3 axes when I did that.

**Reply by Volksolive** [reply]
Got it !

I'll then continue with X Y Z M3 M4 Mn

**Reply by Volksolive** [reply]
I have tested the new PIO code and it seems to work fine, unfortunately because of DK customs nonsense I still don't have Phil Barrett's Pico CNC board to test on. @terjeio do you want to do a pull requests on the 'main' branch or a test branch until it is properly tested?

**Reply by terjeio** [reply]
IMO a PR can wait until properly tested as the current code works.

**Reply by Volksolive** [reply]
I have checked the signal on my DSO and they look as good as the current version. I just don’t have a board or Shift registers to test with and might not have one for quite some time. That’s why I am thinking that if we put it on a test branch and somebody is willing to test it, it could be done. But that’s your call.

**Reply by terjeio** [reply]
Open a new issue and add the changed files in a zip? There are only two or is there a board map as well for testing > 4 axes?
You can also open a PR against the main branch, testers can download from that as well. A plus is that the code can be commented on if you do so.

I have made a 32 bit shift register (4 chained 595s on a small single-sided PCB) that can be used for testing. I can upload the KiCad design files for that if of interest.
